File: java/src/test/java/org/taymyr/lagom/elasticsearch/AggregationsIT.java
package org.taymyr.lagom.elasticsearch;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableMap;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.taymyr.lagom.elasticsearch.SampleProduct.Category;
import org.taymyr.lagom.elasticsearch.SampleProduct.Payload;
import org.taymyr.lagom.elasticsearch.SampleProduct.StaticFacets;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkCreate;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkRequest;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkResult;
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex;
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex.Settings;
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndexResult;
import org.taymyr.lagom.elasticsearch.indices.dsl.DataType;
import org.taymyr.lagom.elasticsearch.indices.dsl.Mapping;
import org.taymyr.lagom.elasticsearch.indices.dsl.MappingProperty;
import org.taymyr.lagom.elasticsearch.search.dsl.SearchRequest;
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.CompositeAggregation;
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.FilterAggregation;
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.NestedAggregation;
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.TermsAggregation;
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.TermsAggregation.FieldSpec;
import org.taymyr.lagom.elasticsearch.search.dsl.query.compound.BoolQuery;
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.Match;
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.MatchQuery;
import org.taymyr.lagom.elasticsearch.search.dsl.query.joining.NestedQuery;
import org.taymyr.lagom.elasticsearch.search.dsl.query.joining.NestedQueryBody;
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.Term;
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.TermQuery;
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.TermsQuery;

import java.util.Date;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.taymyr.lagom.elasticsearch.deser.ServiceCallKt.invokeT;
import static org.taymyr.lagom.elasticsearch.indices.dsl.MappingProperty.mapping;

import static java.lang.Thread.sleep;
import static java.util.Arrays.asList;

class AggregationsIT extends AbstractElasticsearchIT {

    @Test
    @DisplayName("Aggregations should be correct")
    void shouldBeCorrect() throws InterruptedException, ExecutionException, TimeoutException {
        createIndex();
        addItems();
        search();
    }

    void createIndex() throws InterruptedException, ExecutionException, TimeoutException {
        CreateIndex request = new CreateIndex(
                new Settings(1, 1),
                ImmutableMap.of(
                        "product", new Mapping(ImmutableMap.<String, MappingProperty>builder()
                                .put("id", MappingProperty.LONG)
                                .put("category.id", MappingProperty.LONG)
                                .put("category.title", MappingProperty.KEYWORD)
                                .put("fullTextBoosted", mapping().type(DataType.TEXT).analyzer("russian").build())
                                .put("fullText", mapping().type(DataType.TEXT).analyzer("russian").build())
                                .put("sellerId", MappingProperty.LONG)
                                .put("categoryIds", MappingProperty.LONG)
                                .put("updateDate", MappingProperty.DATE)
                                .put("basePrice", MappingProperty.INTEGER)
                                .put("staticFacets", mapping().type(DataType.NESTED).properties(ImmutableMap.of(
                                        "name", MappingProperty.KEYWORD,
                                        "value", MappingProperty.KEYWORD
                                )).build())
                                .put("payload", MappingProperty.OBJECT)
                                .build()
                        ))
        );
        CreateIndexResult result = eventually(elasticIndices.create("product").invoke(request));
        assertThat(result.getIndex()).isEqualTo("product");
        assertThat(result.getAcknowledged()).isTrue();
        assertThat(result.getShardsAcknowledged()).isTrue();
    }

    void addItems() throws InterruptedException, ExecutionException, TimeoutException {
        IndexedSampleProduct testEntity = new IndexedSampleProduct(new SampleProduct(
                13,
                new Category(100, "Ягоды"),
                "Клубника вкусная",
                "Россия Ягоды Фрукты",
                0,
                asList(100L, 200L, 600L),
                new Date(),
                300500,
                asList(
                        new StaticFacets("country", "Аргентина"),
                        new StaticFacets("salemethod", "Упаковка"),
                        new StaticFacets("saleregion", "Ленинградская область"),
                        new StaticFacets("saleregion", "Мурманская область")
                ),
                null,
                new Payload("")
        ));
        IndexedSampleProduct testEntity2 = new IndexedSampleProduct(new SampleProduct(
                11,
                new Category(200, "Овощи"),
                "Помидоры красные",
                "Россия Овощи",
                0,
                asList(100L, 200L, 500L),
                new Date(),
                200500,
                asList(
                        new StaticFacets("country", "Украина"),
                        new StaticFacets("salemethod", "Поштучно"),
                        new StaticFacets("saleregion", "Ленинградская область"),
                        new StaticFacets("saleregion", "Московская область")
                ),
                null,
                new Payload("")
        ));
        BulkRequest request = BulkRequest.ofCommands(
                new BulkCreate("13", testEntity),
                new BulkCreate("11", testEntity2)
        );
        BulkResult result = eventually(elasticDocument.bulk("product", "product").invoke(request));
        assertThat(result.getErrors()).isFalse();
        assertThat(result.getItems()).hasSize(2);
    }

    void search() throws InterruptedException, TimeoutException, ExecutionException {
        sleep(1000);
        SearchRequest searchRequest = new SearchRequest(
                MatchQuery.ofMatch(new Match() {
                    private Long id = 11L;

                    public Long getId() {
                        return id;
                    }
                }),
                null,
                null,
                ImmutableMap.of(
                        "staticStringFacetsFiltered", FilterAggregation.of(
                                new NestedQuery(
                                        new NestedQueryBody(
                                                "staticFacets",
                                                BoolQuery.boolQuery().filter(
                                                        TermQuery.ofTerm(new Term() {
                                                            @JsonProperty("staticFacets.name")
                                                            private String staticFacetsName = "country";

                                                            public String getStaticFacetsName() {
                                                                return staticFacetsName;
                                                            }
                                                        }),
                                                        TermsQuery.ofTerms(ImmutableMap.of(
                                                                "staticFacets.value", asList("Украина")
                                                        ))
                                                ).build()
                                        )
                                ),
                                ImmutableMap.of(
                                        "staticStringFacets", NestedAggregation.of(
                                                "staticFacets",
                                                ImmutableMap.of(
                                                        "name", TermsAggregation.of(
                                                                new FieldSpec("staticFacets.name"),
                                                                ImmutableMap.of(
                                                                        "value", TermsAggregation.of(new FieldSpec("staticFacets.value"))
                                                                )
                                                        )
                                                )
                                        )
                                )
                        ),
                        "staticStringFacets", NestedAggregation.of(
                                "staticFacets",
                                ImmutableMap.of(
                                        "name", TermsAggregation.of(
                                                new FieldSpec("staticFacets.name"),
                                                ImmutableMap.of(
                                                        "value", TermsAggregation.of(
                                                                new FieldSpec("staticFacets.value")
                                                        )
                                                )
                                        )
                                )
                        ),
                        "categories", new CompositeAggregation(
                                new CompositeAggregation.Composite(asList(
                                        ImmutableMap.of(
                                                "categoryId", new TermsAggregation(new FieldSpec("category.id"))
                                        ),
                                        ImmutableMap.of(
                                                "categoryTitle", new TermsAggregation(new FieldSpec("category.title"))
                                        )
                                ))
                        )
                )
        );
        SampleProductResult result = eventually(invokeT(elasticSearch.search("product", "product"), searchRequest, SampleProductResult.class));
        assertThat(result.getTamedOut()).isFalse();
        SampleCategoryForProduct category = result.getTyped("/aggregations/categories/after_key", SampleCategoryForProduct.class);
        assertThat(category.getCategoryId()).isEqualTo(200);
    }
}

File: src/test/kotlin/org/taymyr/lagom/elasticsearch/AggregationsIT.kt
package org.taymyr.lagom.elasticsearch

import com.fasterxml.jackson.annotation.JsonProperty
import io.kotlintest.extensions.TestListener
import io.kotlintest.matchers.beInstanceOf
import io.kotlintest.matchers.collections.shouldHaveSize
import io.kotlintest.shouldBe
import io.kotlintest.specs.WordSpec
import io.kotlintest.whenReady
import org.taymyr.lagom.elasticsearch.deser.invokeT
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkCreate
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkRequest
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkResult
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex.Mapping
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex.Settings
import org.taymyr.lagom.elasticsearch.indices.dsl.DataType
import org.taymyr.lagom.elasticsearch.indices.dsl.MappingProperty
import org.taymyr.lagom.elasticsearch.search.dsl.SearchRequest
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.CompositeAggregation
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.FilterAggregation
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.NestedAggregation
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.TermsAggregation
import org.taymyr.lagom.elasticsearch.search.dsl.query.aggregation.TermsAggregation.FieldSpec
import org.taymyr.lagom.elasticsearch.search.dsl.query.compound.BoolQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.Match
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.MatchQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.joining.NestedQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.joining.NestedQueryBody
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.Term
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.TermQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.TermsQuery
import java.lang.Thread.sleep
import java.util.Date

class AggregationsIT : WordSpec() {

    override fun listeners(): List<TestListener> = listOf(Companion)

    init {
        "test" should {
            "successfully create new index" {
                val request = CreateIndex(
                    Settings(1, 1),
                    mapOf(
                        "product" to Mapping(mapOf(
                            "id" to MappingProperty.LONG,
                            "category.id" to MappingProperty.LONG,
                            "category.title" to MappingProperty.KEYWORD,
                            "fullTextBoosted" to MappingProperty(DataType.TEXT, "russian"),
                            "fullText" to MappingProperty(DataType.TEXT, "russian"),
                            "sellerId" to MappingProperty.LONG,
                            "categoryIds" to MappingProperty.LONG,
                            "updateDate" to MappingProperty.DATE,
                            "basePrice" to MappingProperty.INTEGER,
                            "staticFacets" to MappingProperty(DataType.NESTED, null, mapOf(
                                "name" to MappingProperty.KEYWORD,
                                "value" to MappingProperty.KEYWORD
                            )),
                            "payload" to MappingProperty.OBJECT
                        ))
                    )
                )
                whenReady(elasticIndices.create("product").invoke(request).toCompletableFuture()) { result ->
                    result.acknowledged shouldBe true
                    result.shardsAcknowledged shouldBe true
                    result.index shouldBe "product"
                }
            }
            "successfully add some items" {
                val testEntity = IndexedSampleProduct(SampleProduct(
                    13,
                    SampleProduct.Category(100, "Ягоды"),
                    "Клубника вкусная",
                    "Россия Ягоды Фрукты",
                    0,
                    listOf(100, 200, 600),
                    Date(),
                    300500,
                    listOf(
                        SampleProduct.StaticFacets("country", "Аргентина"),
                        SampleProduct.StaticFacets("salemethod", "Упаковка"),
                        SampleProduct.StaticFacets("saleregion", "Ленинградская область"),
                        SampleProduct.StaticFacets("saleregion", "Мурманская область")
                    ),
                    null,
                    SampleProduct.Payload("")
                ))
                val testEntity2 = IndexedSampleProduct(SampleProduct(
                    11,
                    SampleProduct.Category(200, "Овощи"),
                    "Помидоры красные",
                    "Россия Овощи",
                    0,
                    listOf(100, 200, 500),
                    Date(),
                    200500,
                    listOf(
                        SampleProduct.StaticFacets("country", "Украина"),
                        SampleProduct.StaticFacets("salemethod", "Поштучно"),
                        SampleProduct.StaticFacets("saleregion", "Ленинградская область"),
                        SampleProduct.StaticFacets("saleregion", "Московская область")
                    ),
                    null,
                    SampleProduct.Payload("")
                ))
                val request = BulkRequest.ofCommands(
                    BulkCreate("13", testEntity),
                    BulkCreate("11", testEntity2)
                )
                whenReady(elasticDocument.bulk("product", "product").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe false
                    result.items shouldHaveSize 2
                }
            }
            "do search" {
                sleep(1000)
                @Suppress("unused")
                val searchRequest = SearchRequest(
                    query = MatchQuery.ofMatch(object : Match {
                        val id: Long = 11
                    }),
                    from = null,
                    size = null,
                    aggs = mapOf(
                        "staticStringFacetsFiltered" to FilterAggregation.of(
                            filter = NestedQuery(
                                NestedQueryBody(
                                    path = "staticFacets",
                                    query = BoolQuery.boolQuery().filter(
                                        TermQuery.ofTerm(object : Term {
                                            @JsonProperty("staticFacets.name")
                                            val staticFacetsName = "country"
                                        }),
                                        TermsQuery.ofTerms(mapOf(
                                            "staticFacets.value" to listOf("Украина")
                                        ))
                                    ).build()
                                )
                            ),
                            aggs = mapOf(
                                "staticStringFacets" to NestedAggregation.of(
                                    path = "staticFacets",
                                    aggs = mapOf(
                                        "name" to TermsAggregation.of(
                                            terms = FieldSpec("staticFacets.name"),
                                            aggs = mapOf(
                                                "value" to TermsAggregation.of(FieldSpec("staticFacets.value"))
                                            )
                                        )
                                    )
                                )
                            )
                        ),
                        "staticStringFacets" to NestedAggregation.of(
                            path = "staticFacets",
                            aggs = mapOf(
                                "name" to TermsAggregation.of(
                                    terms = FieldSpec("staticFacets.name"),
                                    aggs = mapOf(
                                        "value" to TermsAggregation.of(
                                            FieldSpec("staticFacets.value")
                                        )
                                    )
                                )
                            )
                        ),
                        "categories" to CompositeAggregation(
                            CompositeAggregation.Composite(listOf(
                                mapOf(
                                    "categoryId" to TermsAggregation(FieldSpec("category.id"))
                                ),
                                mapOf(
                                    "categoryTitle" to TermsAggregation(FieldSpec("category.title"))
                                )
                            ))
                        )
                    )
                )
                whenReady(elasticSearch.search(listOf("product"), listOf("product"))
                    .invokeT<SearchRequest, SampleProductResult>(searchRequest).toCompletableFuture()) { result ->
                    result.tamedOut shouldBe false
                    val category = result.getTyped("/aggregations/categories/after_key", SampleCategoryForProduct::class.java)
                    category.categoryId shouldBe 200
                }
            }
        }
    }

    companion object : LagomClientAndEmbeddedElastic()
}